### DISCUSSION TITLE
getServerSideProps in custom App

### DISCUSSION THREAD

**Discussion by Giuseppetm** [vercel/next.js#39904] [Help] [answered]
https://nextjs.org/docs/advanced-features/custom-app

![image](https://user-images.githubusercontent.com/52317197/186429931-623ab475-4f0b-4094-a744-55939e870efb.png)

Did anyone found a workaround for this?
I would like to use a `getServerSideProps`
```javascript
export const getServerSideProps: GetServerSideProps = async ({ req, res }) => {
    let acceptlang = req.headers["accept-language"]
    console.log("accept language: ", acceptlang)

    return { props: { 
        acceptlang: acceptlang
    } };
};
```

with a custom App:
```javascript
const App = ({ Component, pageProps }: AppProps) => { ..... }
```

**Comment by gaston-flores** [comment, marked as answer]
Did you give `getInitialProps` a try?

```jsx
// _app.js
import App from "next/app";

function MyApp({ Component, pageProps, acceptlang }) {
  console.log(acceptlang);
  return <Component {...pageProps} />;
}

MyApp.getInitialProps = async (appContext) => {
  const appProps = await App.getInitialProps(appContext);
  const acceptlang = appContext.ctx.req.headers["accept-language"];

  return { ...appProps, acceptlang };
};

export default MyApp;
```

Keep in mind this will apply to *all* pages and disable static generation.

**Comment by icyJoseph** [comment]
Depending on what you are trying to achieve, you might be able to implement a few workarounds, but as it is not supported at the core, each workaround will have drawbacks.

`getInitialProps` to read from the request is tricky, **because when it runs on the client, or for pages that use `getStaticProps`, reading the request object will lead to crashing out.** 

You could also implement `getServerSideProps` in all pages, and collect the value you need from `pageProps` in `_app`. Obviously, this is error prone, and very hard to keep up.

Alternatively, using `middleware`, you could rewrite requests to a sub folder `pages/[lang]/`, where your actual app lives. The rewrite would make it so that, the browser shows `/foo/bar`, but based on the accepted language, you render the contents of `pages/en/foo/bar` for the english language. Then using the router, you can read the value of `acceptlang` as part of the page query.

```ts
import { NextResponse, type NextRequest } from "next/server";

const PUBLIC_FILE = /\.(.*)$/;

export function middleware(req: NextRequest) {
  const { pathname } = req.nextUrl;
  if (
    pathname.startsWith("/_next") || // exclude Next.js internals
    pathname.startsWith("/api") || //  exclude all API routes
    pathname.startsWith("/static") || // exclude static files
    PUBLIC_FILE.test(pathname)
  )
    return NextResponse.next();

  const lang = req.headers.get("accept-language");

  req.nextUrl.pathname = `${lang}${pathname}`;
  return NextResponse.rewrite(req.nextUrl);
}
```

```tsx
// pages/[lang]/index.tsx
import type { NextPage } from "next";
import Link from "next/link";
import { useRouter } from "next/router";

const Home: NextPage = () => {
  console.log(useRouter().query); // {}, and then {lang: 'something'}

  return (
    <div>
      Home
    </div>
  );
};

export default Home;
```

Then you can read the `lang` value from the router. Do remember that Next.js statically renders when possible, and for that case, the query object is empty at first render. https://nextjs.org/docs/routing/dynamic-routes#caveats:

> Pages that are statically optimized by [Automatic Static Optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization) will be hydrated without their route parameters provided, i.e query will be an empty object ({}).

If you want to disable that:

```tsx
import "../styles/globals.css";
import type { AppProps } from "next/app";

function MyApp({ Component, pageProps }: AppProps) {
  return <Component {...pageProps} />;
}

MyApp.getInitialProps = () => {
  return {};
};

export default MyApp;
```

Still though, good to remember that at build time pages that use `getStaticProps` are still pre-rendered.

> Warning: You have opted-out of Automatic Static Optimization due to `getInitialProps` in `pages/_app`. This does not opt-out pages with `getStaticProps`

The trade-off with this one is that is involves more moving parts, and a specific file structure on the pages document.

**Comment by Giuseppetm** [comment]
Thank you for the answers;
I will probably use a `getInitialProps` since my application basically needs server side rendering on all the pages.
